### PR TITLE
Feature/disable stat for loadouts

### LINKED ISSUE #4559: Loadout Optimizer: Ignore total tier
https://twitter.com/iihavetoes/status/1187104989715390464

This user is interested in maximizing specific stats, even if it doesn't yield the maximum total tier. One idea would be to add a filtering option for ignoring total tier.

### PR DESCRIPTION
**Pull request by ryan-rushton**
Fixes #4559

### CODE REVIEW

File: src/app/loadout-builder/types.ts [hunk 1]
```diff
@@ -9,6 +9,9 @@ export type StatTypes =
   | 'Discipline'
   | 'Intellect'
   | 'Strength';
+
+export type StatTypesWithTotal = StatTypes | 'Total';
```
Comment (ryan-rushton) [inline src/app/loadout-builder/types.ts]: I was tempted to turn all these string values into enums so I can use them in the other files. I left it as I kept the convention but I think it might be nicer to do this. Let me know if so.

File: config/i18n.json [hunk 2]
```diff
@@ -503,6 +503,8 @@
     "SelectPower": "Minimum Power",
     "SelectPowerDescription": "Sets a minimum power level for the entire loadout (not individual pieces)",
     "SelectedPerkKey": "Selected perk",
+    "StatToggleIgnore": "Click this stat to ignore it.",
```
Comment (ryan-rushton) [inline config/i18n.json]: Wasn't sure if should just use include/ignore or two translations as I have. Happy to change.

File: src/app/loadout-builder/LoadoutBuilder.tsx [hunk 3]
```diff
@@ -135,7 +136,16 @@ export class LoadoutBuilder extends React.Component<Props & UIViewInjectedProps,
       },
       minimumPower: 750,
       query: '',
-      statOrder: statKeys
+      statOrder: statKeys,
+      enabledStats: new Set([
+        'Total',
```
Comment (ryan-rushton) [inline src/app/loadout-builder/LoadoutBuilder.tsx]: As per other comment these might be nicer as enums, thoughts?

**Comment by bhollis** [comment]
Do you have screenshots / a description of what exactly this changes about the interaction?

**Comment by ryan-rushton** [comment]
Sorry, I should have known to do that.

So starting from the normal loadout builder screen we would expect something like this.

Screen Shot 2019-10-29 at 5 09 32 pm

Now, I don't care about my total tier value. So I click the T23, (in the red box above) a single time with a left click and I get the following.

Screen Shot 2019-10-29 at 5 10 15 pm

Note that the T* for the total is greyed out (or has it's opacity set to 0.4) and that the sort order now has a build with the highest possible mobility first.

Similarly if we were to start from the following and click in the highlighted red section

Screen Shot 2019-10-29 at 5 25 08 pm

We would get the following. Note that the Mobility section is greyed out and the build with the highest Total, then resilience is now at the top.

Screen Shot 2019-10-29 at 5 10 39 pm

Finally if you deselect everything you still get builds but I have no idea how they are ordered.

Basically it lets you click any stat and it will disable it from being used in the sorting of loadouts in the optimiser.

**Comment by bhollis** [comment]
Ah interesting! Is the reason you went with clicking the stats on the builds because Total isn't shown on the stats on the sidebar?

**Comment by ryan-rushton** [comment]
Yeah, as per our discussion I went with the way that adds nothing to the UI as this will only be used by power users. I have been thinking of a way to make it obvious that it's clickable but haven't come up with anything good yet. You get a pointer when hovering over, but that's not always obvious (to me at least).

**Comment by inexorableAce** [comment]
I would never think to click these things, and if I did click them I would not expect them to do this... sort maybe is what I’d expect and based on what happens I may also assume that’s what happened? 

It needs some feedback to let the user know what’s up. But also I’d say we could just add an “ignore” in the min / max drop downs to ignore stats and that would be a lot more inline with how it works today and easier to understand.

**Comment by ryan-rushton** [comment]
Yeah I think that would be an easy switch over. Basically move the click functionality to the drop and drag stat sections, like you said, maybe an ignore option in both drop downs that would disable the whole stat is selected?

Ignore the idea of disabling the total sort and let people filter down the list as much as they need?

Finally I think the total tier display value would need to change as stats are added or removed to make the sorting obvious. Does this sound reasonable or would it be too jarring?

**Comment by kyleshay** [comment]
What if the dropdowns have an `-ignore/exclude-` item? (if one is selected in min/max, the other sets to `ignore` as well. Ignored stats should auto-sort to the bottom and gray in the list.

**Comment by inexorableAce** [comment]
Well if they just wanted in stat that’s true. They want multiples at the expense of others. So today if you set your mob to max your recov selections get small. So you have to make a bunch of random selections to find the “best” combo of mob and recov. Or set your min - maxes wide and comb through 100 stat mixes. 

I think really this ultimately stems from having the min maxes in general instead of a basic selector like we have for perks. If you could just select a stat (or 3) to prioritize and let the tool do the rest that would also work. It’s less finite but also more useful at the end of day.

**Comment by kyleshay** [comment]
To me, this all just sounds like the minimum dropdown isn't being utilized. By default we leave the min all at 0, because technically an overall higher total tier... can be better. If you care to maximize a stat, then set a minimum on it (but the total tier may drop.)

It's very power user focused, but that was the original intent of it. For users to really dive in and figure out what they want. 

Maybe we should have a simpler 'mode' that is the default. (as well as thinking about how to make the power mode more obvious)

**Comment by inexorableAce** [comment]
Yeah the issue it just that it’s hard to find the best total excluding a stat. 

It’s easy to replicate and see the frustration.
-Go to the optimizer
-drag two or more stats that you do care about
-drag one or more to the bottom that you don’t care about
-set your preferences 

Now find the build that has the MOST of the two or more stats you care about. 

You’ll see that’s not quite as easy as it seems. You either have to set the minimums of the stats you do care about many times to find the combination where they even out... or scroll through all the stat combos that were generated. 

That’s why I think ignoring the total isn’t the problem. You want a high total just the highest total where you don’t care how x favors into that total. So I guess it’s more like weighting the stats and how they apply to the total.

**Comment by ryan-rushton** [comment]
> What if the dropdowns have an `-ignore/exclude-` item? (if one is selected in min/max, the other sets to `ignore` as well.) Ignored stats should auto-sort to the bottom and gray in the list.

@kyleshay I think this would provide a good enough approximation for the functionality required. In this scenario, we would ignore the option to handle the total stat, add in the ignore option to all the dropdowns, change the UI to indicate that it is no being used in the sorting. I have 80% of this functionality done and its just a matter of shifting some stuff around.

Otherwise I agree with @inexorableAce that this page is quite hard to get the results I am after. I have been refraining from putting forward too many dramatic ideas here as I am new and I also don't want to seem like I am coming here to modify DIM for my own usage, but I feel like some another perspective on the page might be useful as I have seen a few comments that people struggle with it.

I am going to set up my case here (well at least try) with some context and thoughts. First I'm trying to take my engineer hat off and put my user thoughts first. I have been a DIM user for l long time now (I think 3-4 years) and this is the first time I have felt like I needed the optimiser because stats are sufficiently complex, thus I have been a user of this page for about two weeks. 

I see this page as a search engine for builds. As in I tweak some parameters and (hopefully) what I want rises to the top. Now in most search engines I have used (property sites are what comes to my mind) when I load the page I am used to seeing a section for filtering and a section for sorting. My usual workflow for using this would be the following. Start search (pick the suburbs I want to look in, yes this is just a filter), load results and add more filters (set number of bedrooms ect),  then I sort the results by price or whatever I care about at the time.

The problem I see with the optimiser page is that the filtering and sorting is all bundled up into one section and it is not clear that sorting is an option. The drag and drop is the only sort option available and it sits under a text input with the word filter in it. I had no idea I could sort until I raised an issue. How I would expect to use this page (think build search engine) would be the following.

1. Open page and select character (first filter)
1. Narrow search further by selecting specific armour piece or a minimum for a stat.
1. Set my sorting preferences so that it orders them in the way I want. 

I would expect the sorting sections to have a couple of options and at a minimum I would expect to be able to select a stat (total or other) to be what sorts the list ignoring all other stats. The fact we have an ordering for stat sort is excellent, it's just that the functionality I expect from all the experience I have using search engines is not there. Separating these two things would make the page much clearer IMO.

**Comment by ryan-rushton** [comment]
Ok. Base on the discussion we had I have moved the enabling/disabling to the tier select drop downs. Heres a couple of screen shots to show the workflow/effects. Let me know thoughts please people.

This is the ignore option (which is translatable)

Screen Shot 2019-11-02 at 10 20 05 pm

On selecting ignore this happens

Screen Shot 2019-11-02 at 10 20 20 pm

On selecting the drop down again you are presented with the include option

Screen Shot 2019-11-02 at 10 21 11 pm

On selecting include it will remember last selected your options. Alternatively a tier can be selected which will select the tier for min/max and the max/min will use the last known value.

If all things are ignored this happens

Screen Shot 2019-11-02 at 10 20 43 pm

File: src/app/loadout-builder/generated-sets/TierSelect.tsx [hunk 4]
```diff
@@ -135,44 +140,61 @@ function MinMaxSelectInner({
...
   const value = type === 'Min' ? Math.max(min, stats[stat].min) : Math.min(max, stats[stat].max);
-
   return (
-    <select value={value} onChange={handleChange}>
+    <select value={ignored ? '-' : value} onChange={handleChange}>
       <option disabled={true}>{t(`LoadoutBuilder.Select${type}`)}</option>
-      {/*
-        t('LoadoutBuilder.SelectMin')
-        t('LoadoutBuilder.SelectMax')
-       */}
```
Comment (delphiactual) [inline src/app/loadout-builder/generated-sets/TierSelect.tsx]: put this back

Comment (ryan-rushton) [reply src/app/loadout-builder/generated-sets/TierSelect.tsx]: Done, is it just to point out what the translations above can be?

**Comment by ryan-rushton** [comment]
Thanks @inexorableAce, I was going to ask if we wanted to go ahead with this or not this week. I will fix the conflicts later today. Also I agree, it would be good to give it a week in beta.

**Comment by bhollis** [comment]
Sorry, I lost track of this. Let's at least let people play with it over the weekend.
